cassiopeia/tools/branch_length_estimator.py

```diff
@@ -0,0 +1,325 @@
+        for (parent, child) in tree.edges():
+            edge_length = r_X_t_variables[parent] - r_X_t_variables[child]
+            # TODO: hardcoded '0' here...
+            zeros_parent = tree.get_state(parent).count("0")
+            zeros_child = tree.get_state(child).count("0")
+            new_cuts_child = zeros_parent - zeros_child
+            assert new_cuts_child >= 0
+            # Add log-lik for characters that didn't get cut
+            log_likelihood += zeros_child * (-edge_length)
+            # Add log-lik for characters that got cut
+            log_likelihood += new_cuts_child * cp.log(1 - cp.exp(-edge_length))
+
+        # # # # # Add regularization # # # # #
+
+        l2_penalty = 0
+        for (parent, child) in tree.edges():
+            for child_of_child in tree.children(child):
+                edge_length_above = (
+                    r_X_t_variables[parent] - r_X_t_variables[child]
+                )
+                edge_length_below = (
+                    r_X_t_variables[child] - r_X_t_variables[child_of_child]
+                )
+                l2_penalty += (edge_length_above - edge_length_below) ** 2
+        l2_penalty *= l2_regularization
+
+        # # # # # Solve the problem # # # # #
+
+        obj = cp.Maximize(log_likelihood - l2_penalty)
+        prob = cp.Problem(obj, all_constraints)
+
+        f_star = prob.solve(solver="ECOS", verbose=verbose)
+
+        # # # # # Populate the tree with the estimated branch lengths # # # # #
+
+        for node in tree.nodes():
+            tree.set_age(node, age=r_X_t_variables[node].value)
+
+        for (parent, child) in tree.edges():
+            new_edge_length = (
+                r_X_t_variables[parent].value - r_X_t_variables[child].value
+            )
+            tree.set_edge_length(parent, child, length=new_edge_length)
+
+        self.log_likelihood = log_likelihood.value
+        self.log_loss = f_star
+
+    @classmethod
+    def log_likelihood(self, tree: Tree) -> float:
+        r"""
+        The log-likelihood of the given tree under the model.
+        """
+        log_likelihood = 0.0
+        for (parent, child) in tree.edges():
+            edge_length = tree.get_age(parent) - tree.get_age(child)
+            # TODO: hardcoded '0' here...
+            zeros_parent = tree.get_state(parent).count("0")
+            zeros_child = tree.get_state(child).count("0")
+            new_cuts_child = zeros_parent - zeros_child
+            assert new_cuts_child >= 0
+            # Add log-lik for characters that didn't get cut
+            log_likelihood += zeros_child * (-edge_length)
+            # Add log-lik for characters that got cut
+            if edge_length < 1e-8 and new_cuts_child > 0:
+                return -np.inf
+            log_likelihood += new_cuts_child * np.log(1 - np.exp(-edge_length))
+        return log_likelihood
+
+
+class IIDExponentialBLEGridSearchCV(BranchLengthEstimator):
```

Comment:
Sounds good!

I used to follow your pattern until very recently, only because I started to read about people's thought on this and it seems people _tend_ to prefer the several-related-classes-per-module way:


***
(No need to read these, just here for reference)
https://softwareengineering.stackexchange.com/questions/209982/is-it-considered-pythonic-to-have-multiple-classes-defined-in-the-same-file
https://stackoverflow.com/questions/106896/how-many-classes-should-i-put-in-one-file
https://www.reddit.com/r/Python/comments/41haw4/why_do_python_style_conventions_prefer_multiple/
***

However, our classes have very small interfaces and large implementations, so it does seem like the one-file-per-class you propose would be more suitable? Citing one of the answers above:

_If your concrete implementations have very different internal concerns, your single file accumulates all those concerns. For example, implementations with non-overlapping dependencies make your single file depend on the union of all those dependencies._

_So, it might sometimes be reasonable to consider the sub-classes' coupling to their dependencies outweighs their coupling to the interface (or conversely, the concern of implementing an interface is weaker than the concerns internal to that implementation)._

_As a specific example, take a generic database interface. Concrete implementations using an in-memory DB, an SQL RDBMS and a web query respectively may have nothing in common apart from the interface, and forcing everyone who wants the lightweight in-memory version to also import an SQL library is nasty._

What do you think about having a `tools/branch_length_estimation` package with one file per class? (In this case `BranchLengthEstimator.py`, `IIDExponentialBLE.py`, `IIDExponentialBLEGridSearchCV.py`).

Reply:
I see the advantage of using the several-classes-per-file if the API and dependencies are all very similar, but yes I feel that the models we'll be developing with the BranchLengthEstimator subclasses will all vary quite a bit in their implementations. For example, the bayesian estimators, I predict, will require quite a bit of extra implementation that's not overlapping with the maximum likelihood based estimators.

So, maybe it would be worthwhile to keep everything separate. I don't think this will become unwieldy in the future, and in my opinion helps readability and interpretability of the codebase.